### ISSUE TITLE
build: configure fails for libfabric with NVHPC compiler

### ISSUE THREAD

**Issue by louisespellacy-arm** [pmodels/mpich#6221]
Hello my name is Louise Spellacy, and I work for Arm.

We have observed a issue with building MPICH 4.0.2 with the NVIDIA SDK compiler.
`-Wno-missing-field-initializers` is not a supported switch by nvhpc compiler.

With the configure line:
```
Configuring MPICH version 4.0.2 with  '--enable-debug' '--enable-debuginfo' '--enable-shared' \
'CFLAGS=-fPIC' 'CXXFLAGS=-fPIC' 'FCFLAGS=-fPIC' 'FFLAGS=-fPIC' 'F90=' \
'--prefix=install_dir' 'CC=nvc' 'LDFLAGS=' 'LIBS=' 'CPPFLAGS=' 'CPP=cpp' \
'CXX=nvc++' 'FC=nvfortran' 'F77=nvfortran'
```

The configure build fails on the step:
```
config.lt: creating libtool
checking for dlopen in -ldl... no
configure: error: dlopen not found.  libfabric requires libdl.
configure: error: libfabric configure failed
```

When you observe the `config.log` for `libfabric` you can see the steps are failing on:
```
config.lt:627: creating libtool
configure:14851: checking for dlopen in -ldl
configure:14876: /home/louspe01/.conan/data/nvhpc_cuda_11.7/22.5/warwci/precommit/package/ebde57ce6584c7a8875ca4a3abb4f224a693bd28/Linux_aarch64/22.5/compilers/bin/nvc -o conftest -g -O0 -Wall -Wundef -Wpointer-arith -Wextra -Wno-unused-parameter -Wno-sign-compare -Wno-missing-field-initializers  -fPIC  -fvisibility=hidden    conftest.c -ldl   >&5
nvc-Error-Unknown switch: -Wno-missing-field-initializers
configure:14876: $? = 1
```

To complete the configure step - you need to add `-noswitcherror` to CFLAGS, FCFLAGS, CXXFLAGS and FFLAGS.
This ignores the fact that unsupported switches are reported from the compiler.

From the output, with `-noswitcherror` it looks like there maybe one or two other switch errors further in the libfabric configure step when this one is resolved.

This was using Nvidia SDK 22.5 on sles 15.3

**Comment by hzhou** [comment]
@louisespellacy-arm What if we pass `--disable-debug` to the configure line?

**Comment by louisespellacy-arm** [comment]
If I replace `--enable-debug` with `--disable-debug` the issue doesn't appear.

**Comment by hzhou** [comment]
We could pass `--disable-debug` to libfabric by default, and turn it on under `--enable-strict` or `--enable-g=debug`. That seems more inline with the main mpich configure.

**Comment by louisespellacy-arm** [comment]
Yes that sounds like a possible solution! What would the impact been for having now debug for libfabric - or is this similar approach used for other dependencies?

**Comment by raffenet** [comment]
I created a PR on the libfabric repo to treat `-Wno-missing-field-initializers` differently during configure. Now the script should test if it is accepted by the compiler before adding it to `CFLAGS`.

**Comment by louisespellacy-arm** [comment]
@raffenet That sounds like a better approach as I think there are other switches that are not supported but this is just the first instance that fails.

**Comment by louisespellacy-arm** [comment]
@raffenet What version will your change be available in?

**Comment by raffenet** [comment]
I just picked the upstream commit into our embedded submodule. The fix didn't make 4.0.3, but we should be able to include in 4.1b1 which should be out either this week or next.
